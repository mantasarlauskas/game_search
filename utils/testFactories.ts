import { Category } from 'utils/types';

export function getCategories(): Category[] {
    return [
        {
            id: 18893,
            name: 'Feral Interactive',
            slug: 'feral-interactive',
            games_count: 107,
            image_background: 'https://media.rawg.io/media/games/157/15742f2f67eacff546738e1ab5c19d20.jpg',
            games: [
                {
                    slug: 'rise-of-the-tomb-raider',
                    name: 'Rise of the Tomb Raider',
                },
                {
                    slug: 'shadow-of-mordor',
                    name: 'Middle-earth: Shadow of Mordor',
                },
                {
                    slug: 'hitman',
                    name: 'Hitman',
                },
                {
                    slug: 'bioshock-2',
                    name: 'BioShock 2',
                },
                {
                    slug: 'bioshock-remastered',
                    name: 'BioShock Remastered',
                },
                {
                    slug: 'xcom-enemy-unknown',
                    name: 'XCOM: Enemy Unknown',
                },
            ],
        },
        {
            id: 405,
            name: 'Ubisoft',
            slug: 'ubisoft',
            games_count: 297,
            image_background: 'https://media.rawg.io/media/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg',
            games: [
                {
                    slug: 'far-cry-3',
                    name: 'Far Cry 3',
                },
                {
                    slug: 'for-honor',
                    name: 'For Honor',
                },
                {
                    slug: 'watch-dogs',
                    name: 'Watch Dogs',
                },
                {
                    slug: 'far-cry-5',
                    name: 'Far Cry 5',
                },
                {
                    slug: 'assassins-creed-origins',
                    name: 'Assassin\'s Creed Origins',
                },
                {
                    slug: 'rayman-legends',
                    name: 'Rayman Legends',
                },
            ],
        },
    ];
}
