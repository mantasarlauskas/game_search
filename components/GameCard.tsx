import { useEffect, useRef, useState, MouseEvent } from 'react';
import debounce from 'debounce';
import Link from 'next/link';
import Image from 'next/image';
import styles from 'components/GameCard.module.scss';
import { Game } from 'utils/types';
import InfoList from 'components/InfoList';
import { Route } from 'utils/routes';
import DivButton from 'components/DivButton';
import { getCroppedImageUrl } from 'utils/media';
import { roundNumber } from 'utils/number';

function GameCard({
    game: { slug, name, background_image, clip, platforms, rating },
}: GameCardProps) {
    const imageUrl = getCroppedImageUrl(background_image);
    const [showVideo, setShowVideo] = useState(false);
    const imageRef = useRef<HTMLImageElement>(null);
    const [mediaHeight, setMediaHeight] = useState('auto');

    useEffect(() => {
        const handleResize = debounce(() => {
            const height = imageRef.current?.getBoundingClientRect()?.height;
            setMediaHeight(height ? `${height}px` : 'auto');
        }, 10);

        window.addEventListener('resize', handleResize);
        return () => {
            window.removeEventListener('resize', handleResize);
        };
    }, []);

    function startVideo() {
        setShowVideo(true);
        const height = imageRef.current?.getBoundingClientRect()?.height;
        setMediaHeight(height ? `${height}px` : 'auto');
    }

    function onMouseLeave() {
        setShowVideo(false);
    }

    function onMediaClick(e: MouseEvent) {
        if (clip?.clip) {
            e.stopPropagation();

            if (!showVideo) {
                startVideo();
            }
        }
    }

    return (
        <Link href={`${Route.GAMES}/${slug}`}>
            <div
                onMouseEnter={startVideo}
                onMouseLeave={onMouseLeave}
                className={styles.root}
            >
                <DivButton
                    className={styles.media}
                    style={{ height: mediaHeight }}
                    onClick={onMediaClick}
                >
                    {showVideo && clip?.clip ? (
                        <video
                            className={styles.video}
                            src={clip.clip}
                            controls
                            loop
                            playsInline
                            autoPlay
                            muted
                        />
                    ) : (
                        imageUrl && (
                            <div ref={imageRef}>
                                <Image
                                    className={styles.image}
                                    height={400}
                                    width={600}
                                    layout="responsive"
                                    src={imageUrl}
                                    alt={name}
                                    quality={50}
                                />
                            </div>
                        )
                    )}
                </DivButton>
                <div className={styles.content}>
                    <div className={styles.header}>
                        <div className={styles.title}>{name}</div>
                        {!!rating && (
                            <div className={styles.score}>
                                {roundNumber(rating)}
                            </div>
                        )}
                    </div>
                    {!!platforms?.length && (
                        <div className={styles.platforms}>
                            {'Platforms: '}
                            <InfoList
                                route={Route.PLATFORMS}
                                list={platforms.map(({ platform }) => platform)}
                            />
                        </div>
                    )}
                </div>
            </div>
        </Link>
    );
}

interface GameCardProps {
    game: Game;
}

export default GameCard;
