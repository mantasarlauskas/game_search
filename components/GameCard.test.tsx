import GameCard from 'components/GameCard';
import { getGames } from 'testing/testFactories';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('<GameCard />', () => {
    const [game] = getGames();

    it('matches snapshot', () => {
        const { container } = render(<GameCard game={game} />);
        expect(container).toMatchSnapshot();
    });

    it('shows game info', () => {
        render(<GameCard game={game} />);
        expect(screen.getByAltText('Portal 2')).toBeInTheDocument();
        expect(screen.getByText('95')).toBeInTheDocument();
        expect(screen.getByText(/PC/)).toBeInTheDocument();
        expect(screen.getByText('Xbox One')).toBeInTheDocument();
        expect(screen.getByText('Portal 2')).toBeInTheDocument();
    });

    it('on image hover shows video', () => {
        const { container } = render(<GameCard game={game} />);
        expect(screen.getByRole('img')).toBeInTheDocument();
        userEvent.hover(screen.getByRole('button'));
        expect(screen.queryByRole('img')).not.toBeInTheDocument();
        expect(container.querySelector('video')).toBeInTheDocument();

        userEvent.unhover(screen.getByRole('button'));
        expect(screen.getByRole('img')).toBeInTheDocument();
        expect(container.querySelector('video')).not.toBeInTheDocument();
    });

    it('on image click shows video', () => {
        const { container } = render(<GameCard game={game} />);
        expect(screen.getByRole('img')).toBeInTheDocument();
        userEvent.click(screen.getByRole('button'));
        expect(screen.queryByRole('img')).not.toBeInTheDocument();
        expect(container.querySelector('video')).toBeInTheDocument();
    });

    it('on name click goes to game page', () => {
        render(<GameCard game={game} />);
        userEvent.click(screen.getByText('Portal 2'));
        expect(window.location.href).toEqual('/games/portal-2');
    });

    it('on platform name click goes to platform page', () => {
        render(<GameCard game={game} />);
        userEvent.click(screen.getByText('Xbox One'));
        expect(window.location.href).toEqual('/platforms/1');
    });
});
