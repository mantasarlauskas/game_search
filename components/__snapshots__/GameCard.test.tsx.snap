// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameCard /> matches snapshot 1`] = `
<div>
  <div>
    <div
      class="root"
    >
      <div
        class="media"
        role="button"
        style="height: auto;"
        tabindex="0"
      >
        <div>
          <img
            alt="Portal 2"
            class="image"
            height="400"
            layout="responsive"
            quality="50"
            src="https://media.rawg.io/media/crop/600/400/games/328/3283617cb7d75d67257fc58339188742.jpg"
            width="600"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class="header"
        >
          <div
            class="title"
          >
            Portal 2
          </div>
          <div
            class="score"
          >
            4.6
          </div>
        </div>
        <div
          class="platforms"
        >
          Platforms: 
          <div>
            <span
              class="item"
            >
              PC, 
            </span>
          </div>
          <div>
            <span
              class="item"
            >
              Xbox One
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
